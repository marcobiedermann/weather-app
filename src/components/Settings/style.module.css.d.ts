declare const styles: {
  readonly form__field: string;
  readonly form__input: string;
  readonly 'form__input--submit': string;
};
export = styles;
