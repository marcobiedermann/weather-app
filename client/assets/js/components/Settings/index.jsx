import { Field, Form, withFormik } from 'formik';
import PropTypes from 'prop-types';
import React from 'react';
import Label from '../Label';
import { languages, units } from '../../constants/localization';

const Settings = props => {
  const { isSubmitting } = props;

  return (
    <Form>
      <div>
        <Label htmlFor="language">Language</Label>
        <Field component="select" name="language">
          {languages.map(language => (
            <option value={language.id} key={language.id}>
              {language.name}
            </option>
          ))}
        </Field>
      </div>
      <div>
        <Label htmlFor="unit">Unit</Label>
        <Field component="select" name="unit">
          {units.map(unit => (
            <option value={unit.id} key={unit.id}>
              {unit.name}
            </option>
          ))}
        </Field>
      </div>
      <div>
        <button type="submit" disabled={isSubmitting}>
          Save
        </button>
      </div>
    </Form>
  );
};

Settings.propTypes = {
  isSubmitting: PropTypes.bool,
};

Settings.defaultProps = {
  isSubmitting: false,
};

export default withFormik({
  handleSubmit: (values, { props, setSubmitting }) => {
    props.updateSettings(values);
    setSubmitting(false);
  },
  mapPropsToValues: props => ({
    language: props.language,
    unit: props.unit,
  }),
})(Settings);
