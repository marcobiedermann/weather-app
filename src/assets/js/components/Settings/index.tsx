import classNames from 'classnames';
import { Field, Form, Formik, FormikHelpers } from 'formik';
import React, { FC } from 'react';
import { languages, units } from '../../constants/localization';
import Label from '../Label';
import styles from './style.module.css';

type Unit = 'metric' | 'imperial';

interface Values {
  language: string;
  unit: Unit;
}

export interface SettingsProps {
  initialValues: Values;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  onSubmit: (values: Values, formikHelpers: FormikHelpers<Values>) => void | Promise<any>;
}

const Settings: FC<SettingsProps> = (props) => {
  return (
    <Formik {...props}>
      {({ isSubmitting }) => (
        <Form>
          <div className={styles.form__field}>
            <Label htmlFor="language">Language</Label>
            <Field component="select" name="language" className={styles.form__input}>
              {languages.map((language) => (
                <option value={language.id} key={language.id}>
                  {language.name}
                </option>
              ))}
            </Field>
          </div>
          <div className={styles.form__field}>
            <Label htmlFor="unit">Unit</Label>
            <Field component="select" name="unit" className={styles.form__input}>
              {units.map((unit) => (
                <option value={unit.id} key={unit.id}>
                  {unit.name}
                </option>
              ))}
            </Field>
          </div>
          <div className={styles.form__field}>
            <button
              type="submit"
              disabled={isSubmitting}
              className={classNames(styles.form__input, styles['form__input--submit'])}
            >
              Save
            </button>
          </div>
        </Form>
      )}
    </Formik>
  );
};

export default Settings;
