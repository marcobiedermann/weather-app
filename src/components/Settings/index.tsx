import { zodResolver } from '@hookform/resolvers/zod';
import clsx from 'clsx';
import { useForm } from 'react-hook-form';
import { z } from 'zod';
import {
  Language,
  SupportedLanguage,
  languages,
  supportedLanguages,
  units,
} from '../../constants/localization';
import Label from '../Label';
import styles from './style.module.css';

function isSupportedLanguage(language: Language): boolean {
  return supportedLanguages.includes(language.id as SupportedLanguage);
}

const formDataSchema = z.object({
  language: z.string(),
  unit: z.union([z.literal('metric'), z.literal('imperial')]),
});

type FormData = z.infer<typeof formDataSchema>;

interface SettingsProps {
  defaultValues: FormData;
  onSubmit: (data: FormData) => void;
}

function Settings(props: SettingsProps): JSX.Element {
  const { defaultValues, onSubmit } = props;
  const { register, handleSubmit } = useForm<FormData>({
    defaultValues,
    resolver: zodResolver(formDataSchema),
  });

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className={styles.form__field}>
        <Label htmlFor="language">Language</Label>
        <select {...register('language')} className={styles.form__input}>
          {languages.filter(isSupportedLanguage).map((language) => (
            <option value={language.id} key={language.id}>
              {language.name}
            </option>
          ))}
        </select>
      </div>
      <div className={styles.form__field}>
        <Label htmlFor="unit">Unit</Label>
        <select {...register('unit')} className={styles.form__input}>
          {units.map((unit) => (
            <option value={unit.id} key={unit.id}>
              {unit.name}
            </option>
          ))}
        </select>
      </div>
      <div className={styles.form__field}>
        <button type="submit" className={clsx(styles.form__input, styles['form__input--submit'])}>
          Save
        </button>
      </div>
    </form>
  );
}

export type { FormData, SettingsProps };
export default Settings;
