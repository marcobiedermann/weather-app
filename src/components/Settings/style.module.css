.form__field {
  margin-bottom: 1.5em;
}

.form__input {
  padding: 1em 1.5em;
  width: 100%;
}

.form__input--submit {
  width: auto;
}
