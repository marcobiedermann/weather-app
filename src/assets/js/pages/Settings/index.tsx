import { FormikHelpers } from 'formik';
import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { useDispatch, useSelector } from 'react-redux';
import Settings from '../../components/Settings';
import { selectSettings, updateSettings } from '../../reducers/settings';

type Unit = 'metric' | 'imperial';

interface Values {
  language: string;
  unit: Unit;
}

const SettingsPage: FC = () => {
  const dispatch = useDispatch();
  const { language, unit } = useSelector(selectSettings);

  const initialValues: Values = {
    language,
    unit,
  };

  function onSubmit(values: Values, formikHelpers: FormikHelpers<Values>): void | Promise<any> {
    const { setSubmitting } = formikHelpers;
    const { language, unit } = values;

    console.log({ language, unit });
    dispatch(
      updateSettings({
        language,
        unit,
      }),
    );

    setSubmitting(false);
  }

  return (
    <div>
      <Helmet>
        <title>Settings</title>
      </Helmet>
      <Settings initialValues={initialValues} onSubmit={onSubmit} />
    </div>
  );
};

export default SettingsPage;
