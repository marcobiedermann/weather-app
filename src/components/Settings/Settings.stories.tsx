import type { Meta, StoryObj } from '@storybook/react';
import Settings from '.';

const meta: Meta<typeof Settings> = {
  args: {},
  component: Settings,
};

type Story = StoryObj<typeof Settings>;

const Default: Story = {};

export { Default };
export default meta;
