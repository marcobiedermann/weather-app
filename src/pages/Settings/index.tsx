import React from 'react';
import { Helmet } from 'react-helmet';
import { useDispatch, useSelector } from 'react-redux';
import Settings, { FormData } from '../../components/Settings';
import { selectSettings, updateSettings } from '../../reducers/settings';

function SettingsPage(): JSX.Element {
  const dispatch = useDispatch();
  const { language, unit } = useSelector(selectSettings);

  const defaultValues: FormData = {
    language,
    unit,
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  function onSubmit(data: FormData): void | Promise<any> {
    dispatch(updateSettings(data));
  }

  return (
    <>
      <Helmet>
        <title>Settings</title>
      </Helmet>
      <Settings defaultValues={defaultValues} onSubmit={onSubmit} />
    </>
  );
}

export default SettingsPage;
