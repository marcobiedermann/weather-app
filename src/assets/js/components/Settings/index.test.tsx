import React from 'react';
import renderer from 'react-test-renderer';
import Settings from '.';

describe('components/Settings', () => {
  it('renders correctly', () => {
    expect.assertions(1);

    const tree = renderer
      .create(<Settings initialValues={{ language: 'en', unit: 'metric' }} onSubmit={jest.fn()} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
