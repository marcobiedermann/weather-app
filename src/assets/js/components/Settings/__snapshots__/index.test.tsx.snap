// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Settings renders correctly 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="form__field"
    >
      <label
        for="language"
      >
        Language
      </label>
      <select
        class="form__input"
        name="language"
      >
        <option
          value="ar"
        >
          Arabic
        </option>
        <option
          value="bg"
        >
          Bulgarian
        </option>
        <option
          value="ca"
        >
          Catalan
        </option>
        <option
          value="zh"
        >
          Chinese Simplified
        </option>
        <option
          value="hr"
        >
          Croatian
        </option>
        <option
          value="cz"
        >
          Czech
        </option>
        <option
          value="nl"
        >
          Dutch
        </option>
        <option
          value="en"
        >
          English
        </option>
        <option
          value="fi"
        >
          Finnish
        </option>
        <option
          value="fr"
        >
          French
        </option>
        <option
          value="de"
        >
          German
        </option>
        <option
          value="el"
        >
          Greek
        </option>
        <option
          value="gl"
        >
          Galician
        </option>
        <option
          value="hu"
        >
          Hungarian
        </option>
        <option
          value="it"
        >
          Italian
        </option>
        <option
          value="ja"
        >
          Japanese
        </option>
        <option
          value="kr"
        >
          Korean
        </option>
        <option
          value="la"
        >
          Latvian
        </option>
        <option
          value="lt"
        >
          Lithuanian
        </option>
        <option
          value="mk"
        >
          Macedonian
        </option>
        <option
          value="fa"
        >
          Persian (Farsi)
        </option>
        <option
          value="pl"
        >
          Polish
        </option>
        <option
          value="pt"
        >
          Portuguese
        </option>
        <option
          value="ro"
        >
          Romanian
        </option>
        <option
          value="ru"
        >
          Russian
        </option>
        <option
          value="sk"
        >
          Slovak
        </option>
        <option
          value="sl"
        >
          Slovenian
        </option>
        <option
          value="es"
        >
          Spanish
        </option>
        <option
          value="se"
        >
          Swedish
        </option>
        <option
          value="tr"
        >
          Turkish
        </option>
        <option
          value="ua"
        >
          Ukrainian
        </option>
        <option
          value="vi"
        >
          Vietnamese
        </option>
      </select>
    </div>
    <div
      class="form__field"
    >
      <label
        for="unit"
      >
        Unit
      </label>
      <select
        class="form__input"
        name="unit"
      >
        <option
          value="imperial"
        >
          imperial
        </option>
        <option
          value="metric"
        >
          metric
        </option>
      </select>
    </div>
    <div
      class="form__field"
    >
      <button
        class="form__input form__input--submit"
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</DocumentFragment>
`;
